### ISSUE TITLE
Faulty scaling logic when using runner groups

### ISSUE THREAD

**Issue by toast-gear** [philips-labs/terraform-aws-github-runner#1370]
👋  I've been getting odd scaling of my runners for a while whilst I've been testing them and I think I've found some issues with the scaling logic.

In my organisation we are using runner groups due to our need for server-side security. The benefit of my 403 woes https://github.com/philips-labs/terraform-aws-github-runner/issues/1348 is that I think I've noticed some issues with the current scaling logic. We don't do any filtering based on runner group as far as I can tell.

When a `workflow_job` webhook is recieved its signiture is validated and the labels are checked to see if they are valid for that stack. If these things pass then the job is added onto the SQS queue. There are a couple of problems with this logic at the moment:

1. If you are running multiple runner groups that means you are probably running multiple webhooks (same problem exists with a single central app webhook too though). You additionally are probably using generic labels as the labels are for workflow resource allocation and to allow workflows to target capabilities, they're a client resource allocation feature rather than a security feature. As a result runner groups will often share labels as the label name isn't a runner group specific thing e.g. `medium`, `default`, `docker` etc etc. This most likely means that every event that passes the webhook checks gets allocated to N number of SQS queues which in turn results in N number of instances being scaled up, where N is the instances of this module you have deployed (1 per runner group).

The runner group needs to be set as an evnrionment variable for the webhook lambda and the runner groups in the payload need to be checked so that only events that are bound for that webhooks runner group are added to that webhooks SQS queue. If the job isn't bound for that webhooks runner group then the job should be ignored by that webhook lambda as it is intended for another instance of the stack which is deployed for that runner group.

2. I don't think this is a problem as it looks like we check to see if the payload is queued in the scale up lambda but I thought I'd mention it. I noticed I got green webhook responses for workflow job events where the job was skipped due to conditions in the workflow.

My events that reported a 200 response and were skipped jobs had these attributes of interest:

```json
{
  "action": "completed",
  "workflow_job": {
    ...
    "status": "completed",
    "conclusion": "skipped",
     ...
    "labels": [

    ],
    "runner_id": null,
    "runner_name": null,
    "runner_group_id": null,
    "runner_group_name": null
  },
...
```

I think the general design approach that should be taken is the webhook lambda is treated like a broker service as it sort of has to be one in terms of filtering when you are running multiple instances to support runner groups. All the required filtering logic should be in the webhook lambda so only jobs that actually need to be handled are added to the apppropriate SQS queue (some defensive code in the scale-up lambda is also good but the onus on correct filtering should be in the webhook lambda).

That's just my 2 pence worth though, resolving the faulty scaling by checking the runner group in the webhook lambda would make me happy :D.

@npalm this is probably going to prevent me from rolling this solution out as intended. We are starting roll out Actions as our defacto CI/CD platform. With the current faulty logic I am going to have to massively reduce the max runner count on all my runner groups resulting in a very non-optimal setup.

**Comment by npalm** [comment]
@toast-gear agree, filtering should be done in the webhook when possible. Do you have the time to propose a solution? Adding a filter the the webhook should not be hard. The support of the workflow_job is realtive new. At the time of the check_run we did not have the options like that

**Comment by npalm** [comment]
Link to github doc for workflow_job https://docs.github.com/en/developers/webhooks-and-events/webhooks/webhook-events-and-payloads#workflow_job

**Comment by npalm** [comment]
Just checked an event. The event sent for a queued job does not contain the runner group. I assume this information will only available once the job is started.

```
    "runner_id": null,
    "runner_name": null,
    "runner_group_id": null,
    "runner_group_name": null
```

**Comment by toast-gear** [comment]
You're spot on, I don't think there is anything you can do module wise. This perhaps is a documentation ticket at this point.

**Comment by toast-gear** [comment]
It took me a disturbing amount of time but I've finally realised how GitHub are expecting you to use runner groups and it isn't the way I thought. I raised a ticket in ARC with how they are supposed to work to improve the docs, thought I might as well link it here seen as it's very related https://github.com/actions-runner-controller/actions-runner-controller/issues/1055. Going to close this ticket as in order for teraform-aws-github-runner to scale runner groups with no redudent scaling a broker would be needed which is covered in https://github.com/philips-labs/terraform-aws-github-runner/issues/1282.
